rfcs/77-dev-tools-abstraction.md

```diff
@@ -0,0 +1,658 @@
+
+### Why not add information about layout to our traits?
+
+This sort of information is really helpful for toolboxes like editors to ensure that dev tools don't overlap each other: allowing them to be laid out beside each other or enabled one at a time when necessary.
+
+However:
+
+1. This RFC is complex enough: keeping its scope small will make it easier to review and implement.
+2. Unlike with a CLI-style interface the correct abstractions aren't immediately obvious: we should prototype more toolkits and tools to explore the needs of the space further.
+3. Not all dev tools need layout information: we would need to use optional methods and/or subtraits to fill this need, adding further complexity.
+
+## Unresolved questions
+
+1. Are the provided methods (along with those on `Reflect`) sufficient to allow toolkits to populate and run a wide range of dev tools without requiring an additional trait and manual registration?
+2. Is the use of `downcast_ref` / `downcast_mut` correct in `DevToolsRegistry::get`?
+3. Are `ModalDevTool` and `DevCommand` the best names?
+4. How, precisely, can we achieve the sort of API shown in [Building toolboxes] using reflection?
+   1. A prototype would be great here.
+5. Does the function pointer approach used by `DevCommandMetadata` compile and work successfully?
+   1. Again, prototype please!
+
+## Future possibilities
+
+Related but out-of-scope questions:
+
+- [where should dev tools live](https://github.com/bevyengine/bevy/pull/12354#issuecomment-1982284605)?
+  - where the primitives used are defined? e.g. bevy_gizmos
+  - where they're used for debugging? e.g. bevy_sprite
+  - in bevy_dev_tools?
+  - in dedicated crates, saving bevy_dev_tools for higher level abstractions?
+- should dev tools be embedded in each application or should testing be done through an editor which controls these dev tools?
+  - this is one of the [key questions](https://github.com/bevyengine/bevy_editor_prototypes/discussions/1) for the bevy_editor efforts
+
+Future possibilities:
+
+1. Over time, we can extend the `ModalDevTool` and `DevCommand` traits with more methods (mandatory and optional), enabling:
+   1. Gradual unification of more complex shared configuration (e.g. a `set_font` method).
+   2. Serializing and deserializing structs from a `.bsn` file, rather than simple CLI-style strings, allowing toolboxes to build abstractions for persistent config.
+2. Add support for a trait-queries-on-resources operation, and stop storing modal dev tools in the `DevToolsRegistry`.
+3. A derive macro for `DevCommand`, that aligns with `clap`'s calling conventions.
+4. Add specialized types of dev tools, likely each with their own trait to establish conventions and resolve conflicts between them.
+    1. For example, we may want to coordinate between various UI elements or overlays to avoid positioning conflicts, dev tools that change the materials of objects, or camera controllers.
```

Comment:
```suggestion
    1. For example, we may want to coordinate between various UI elements or overlays to avoid positioning conflicts, dev tools that change the materials of objects, or camera controllers. See [Discussion #12589](https://github.com/bevyengine/bevy/discussions/12589) for some initial exploration.
```